File: app01_java_block/src/main/java/com/devpowers/legalchain/p2p/PeerSession.java
package com.devpowers.legalchain.p2p;

import org.springframework.web.socket.WebSocketSession;

import javax.crypto.SecretKey;
import java.security.PrivateKey;

/**
 * Mutable handshake state for one peer link. A link becomes {@code established} only
 * after the ML-DSA-authenticated ML-KEM handshake has produced a one-time AES-256
 * session key; until then no ledger data flows.
 */
public class PeerSession {

    private final WebSocketSession ws;
    private final boolean initiator;
    private volatile String peerNodeId;
    private volatile String url;
    /** Initiator's ephemeral ML-KEM private key, discarded once the session key is derived. */
    private volatile PrivateKey ephemeralKemPrivate;
    private volatile SecretKey sessionKey;

    public PeerSession(WebSocketSession ws, boolean initiator) {
        this.ws = ws;
        this.initiator = initiator;
    }

    public WebSocketSession ws() {
        return ws;
    }

    public boolean initiator() {
        return initiator;
    }

    public String peerNodeId() {
        return peerNodeId;
    }

    public void peerNodeId(String peerNodeId) {
        this.peerNodeId = peerNodeId;
    }

    public String url() {
        return url;
    }

    public void url(String url) {
        this.url = url;
    }

    public PrivateKey ephemeralKemPrivate() {
        return ephemeralKemPrivate;
    }

    public void ephemeralKemPrivate(PrivateKey key) {
        this.ephemeralKemPrivate = key;
    }

    public SecretKey sessionKey() {
        return sessionKey;
    }

    public void establish(SecretKey key) {
        this.sessionKey = key;
        this.ephemeralKemPrivate = null; // forward secrecy: ephemeral key is single-use
    }

    public boolean established() {
        return sessionKey != null;
    }
}

File: app03_csharp_block/src/LegalChain/P2p/PeerSession.cs
using System.Net.WebSockets;
using Org.BouncyCastle.Crypto.Parameters;

namespace LegalChain.P2p;

/// <summary>
/// Mutable handshake state for one peer link. A link becomes <see cref="Established"/>
/// only after the ML-DSA-authenticated ML-KEM handshake has produced a one-time AES-256
/// session key; until then no ledger data flows.
/// </summary>
public class PeerSession
{
    public PeerSession(WebSocket socket, bool initiator)
    {
        Socket = socket;
        Initiator = initiator;
    }

    public string Id { get; } = Guid.NewGuid().ToString();

    public WebSocket Socket { get; }

    public bool Initiator { get; }

    public string? PeerNodeId { get; set; }

    public string? Url { get; set; }

    /// <summary>Initiator's ephemeral ML-KEM private key, discarded once the session key is derived.</summary>
    public MLKemPrivateKeyParameters? EphemeralKemPrivate { get; set; }

    public byte[]? SessionKey { get; private set; }

    /// <summary>Serializes concurrent sends: a WebSocket allows only one outstanding send.</summary>
    public SemaphoreSlim SendLock { get; } = new(1, 1);

    public void Establish(byte[] sessionKey)
    {
        SessionKey = sessionKey;
        EphemeralKemPrivate = null; // forward secrecy: ephemeral key is single-use
    }

    public bool Established => SessionKey is not null;
}
